File: examples/code/simple_pcb.jl
#!/usr/bin/env julia
#
# simple_pcb.jl
#    Minimal PCB example for quick testing
#
# usage:
#    cd examples/code
#    julia simple_pcb.jl | julia -t auto ../../frep_multicore.jl [dpi [filename]]
#    OR from examples/:
#    julia code/simple_pcb.jl | python3 ../frep_multicore.py [dpi [filename]]
#

using JSON

# Include the main Fpcb.jl file
include("../../Fpcb.jl")

############################################################
# Simple PCB Example: Minimal board with one LED
############################################################

# Board parameters - smaller board for faster rendering
const width = 0.5       # board width in inches (smaller)
const height = 0.5      # board height in inches (smaller)
const x0 = 0.5          # x origin
const y0 = 0.5          # y origin
const zt = 0.0          # top z
const zb = -0.06        # bottom z
const w = 0.015         # wire width
const mask = 0.004      # solder mask size
const border = 0.05     # image render border

# Create PCB
pcb = PCB(x0, y0, width, height, mask)

# Add just one simple LED (minimal components)
led1 = LED_1206("LED1")
pcb = add_component(pcb, led1, x0 + width/2, y0 + height/2, 0.0, angle=0.0)

# Add simple text label
pcb = add_text(pcb, "Simple PCB", x0 + width / 2.0, y0 + height - 0.08, 
               line=0.01, color_value=UInt32(COLOR_WHITE))

# Generate output - use TOP_FULL for faster rendering (single layer, no bottom)
outputs = generate_output(pcb, x0, y0, zb, zt, border, "TOP_FULL")

# Output JSON
json_str = JSON.json(outputs)

# If no args, emit JSON to stdout for piping into frep.jl
if length(ARGS) == 0
    println(json_str)
elseif length(ARGS) <= 2
    frep_path = joinpath(@__DIR__, "..", "..", "frep_multicore.jl")
    cmd = `julia -t auto $frep_path $(ARGS...)`
    open(cmd, "w") do io
        write(io, json_str)
    end
else
    println("usage: julia simple_pcb.jl [dpi [filename]]")
    exit(1)
end

File: simple_pcb.py
#!/usr/bin/env python3
#
# simple_pcb.py
#    Minimal PCB example — outputs FRep JSON for frep_multicore.py
#
# usage:
#    python3 simple_pcb.py | python3 frep_multicore.py [dpi [filename]]
#    python3 simple_pcb.py | python3 frep_multicore.py --c [dpi [filename]]
#
# A 1" x 1" board with one pad. Same JSON format as pcb.py.

import json
import sys

# ---------------------------------------------------------------------------
# Minimal FRep primitives (string expressions in X, Y, Z)
# ---------------------------------------------------------------------------

def color(c, part):
    return f'({c}*(({part})!=0))'

def add(part1, part2):
    return f'(({part1}) | ({part2}))'

def subtract(part1, part2):
    return f'(({part1}) & ~({part2}))'

def rectangle(x0, x1, y0, y1):
    return f'((X >= ({x0})) & (X <= ({x1})) & (Y >= ({y0})) & (Y <= ({y1})))'

def circle(x0, y0, r):
    return f'(((X-({x0}))*(X-({x0})) + (Y-({y0}))*(Y-({y0}))) <= ({r}*{r}))'

# Colors (RGB packed)
Tan   = (60 << 16) + (90 << 8) + (125 << 0)
White = (255 << 16) + (255 << 8) + (255 << 0)
Blue  = (225 << 8)

# ---------------------------------------------------------------------------
# Board: 1" x 1" with border and one pad
# ---------------------------------------------------------------------------
x, y = 0.0, 0.0
width, height = 1.0, 1.0
border = 0.05

board    = rectangle(x, x + width, y, y + height)
exterior = rectangle(x - border, x + width + border, y - border, y + height + border)
pad      = circle(x + width/2, y + height/2, 0.05)

# One layer (top); output: board (tan) + exterior outline (white) + pad (white)
zt = 0
outputs = {
    "type": "RGB",
    "xmin": x - border,
    "xmax": x + width + border,
    "ymin": y - border,
    "ymax": y + height + border,
    "mm_per_unit": 25.4,
    "layers": [zt],
    "function": add(add(color(Tan, board), color(White, exterior)), color(White, pad)),
}

json.dump(outputs, sys.stdout)
